quickwit-cluster/src/cluster.rs

```diff
@@ -0,0 +1,291 @@
+    artillery_cluster: Arc<ArtilleryCluster>,
+
+    /// A receiver(channel) for exchanging members in a cluster.
+    members: watch::Receiver<Vec<Member>>,
+}
+
+impl Cluster {
+    /// Create a cluster given a host key and a listen address.
+    /// When a cluster is created, the thread that monitors cluster events
+    /// will be started at the same time.
+    pub fn new(host_key: Uuid, listen_addr: SocketAddr) -> anyhow::Result<Self> {
+        info!( host_key=?host_key, listen_addr=?listen_addr, "Create new cluster.");
+        let config = ArtilleryClusterConfig {
+            cluster_key: CLUSTER_ID.as_bytes().to_vec(),
+            listen_addr,
+            ..Default::default()
+        };
+        let (artillery_cluster, _) =
+            ArtilleryCluster::new_cluster(host_key, config).map_err(|err| anyhow::anyhow!(err))?;
+
+        let (members_sender, members_receiver) = watch::channel(Vec::new());
+
+        // Create cluster.
+        let cluster = Cluster {
+            listen_addr,
+            artillery_cluster: Arc::new(artillery_cluster),
+            members: members_receiver,
+        };
+
+        // Add itself as the initial member of the cluster.
+        let member = Member {
+            host_key,
+            listen_addr,
+            is_self: true,
+        };
+        let initial_members: Vec<Member> = vec![member];
+        if let Err(_) = members_sender.send(initial_members) {
+            error!("Failed to add itself as the initial member of the cluster.");
+        }
+
+        // Prepare to start a thread that will monitor cluster events.
+        let thread_inner = cluster.artillery_cluster.clone();
+        let thread_listen_addr = cluster.listen_addr.clone();
+
+        // Start to monitor the cluster events.
+        tokio::task::spawn_blocking(move || {
+            for (artillery_members, artillery_member_event) in thread_inner.events.iter() {
+                log_artillery_event(artillery_member_event);
+                let updated_memberlist: Vec<Member> = artillery_members
+                    .into_iter()
+                    .filter(|member| match member.state() {
+                        ArtilleryMemberState::Alive | ArtilleryMemberState::Suspect => true,
+                        ArtilleryMemberState::Down | ArtilleryMemberState::Left => false,
+                    })
+                    .map(|member| convert_member(member, thread_listen_addr))
+                    .collect();
+                debug!(updated_memberlist=?updated_memberlist);
+                if let Err(_) = members_sender.send(updated_memberlist) {
+                    // Somehow the cluster has been dropped.
+                    break;
+                }
+            }
+        });
+
+        Ok(cluster)
+    }
+
+    pub fn member_change_watcher(&self) -> WatchStream<Vec<Member>> {
+        WatchStream::new(self.members.clone())
+    }
+
+    pub async fn members(&self) -> Vec<Member> {
```

Comment:
remove async (that's my code, my bad)

Reply:
Thanks.

Reply:
Changed.
Please see: https://github.com/quickwit-inc/quickwit/pull/143/commits/cc6acc406a9469092ca23cf4d7bea59acb481db4